File: android/app/src/main/java/info/bagen/rust/plaoc/microService/sys/mwebview/MultiWebViewNMM.kt
package info.bagen.rust.plaoc.microService.sys.mwebview

import android.content.Intent
import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import androidx.compose.runtime.mutableStateMapOf
import info.bagen.rust.plaoc.App
import info.bagen.rust.plaoc.microService.core.AndroidNativeMicroModule
import info.bagen.rust.plaoc.microService.core.BootstrapContext
import info.bagen.rust.plaoc.microService.core.MicroModule
import info.bagen.rust.plaoc.microService.core.NativeMicroModule
import info.bagen.rust.plaoc.microService.helper.Mmid
import info.bagen.rust.plaoc.microService.helper.ioAsyncExceptionHandler
import info.bagen.rust.plaoc.microService.helper.printdebugln
import info.bagen.rust.plaoc.microService.ipc.Ipc
import io.ktor.util.collections.*
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import org.http4k.core.Method
import org.http4k.core.Response
import org.http4k.core.Status
import org.http4k.lens.Query
import org.http4k.lens.string
import org.http4k.routing.bind
import org.http4k.routing.routes
import java.util.concurrent.ConcurrentSkipListSet

inline fun debugMultiWebView(tag: String, msg: Any? = "", err: Throwable? = null) =
    printdebugln("mwebview", tag, msg, err)

class MultiWebViewNMM : AndroidNativeMicroModule("mwebview.sys.dweb") {
    class ActivityClass(var mmid: Mmid, val ctor: Class<out MultiWebViewActivity>)

    companion object {
        val activityClassList = mutableListOf(
            ActivityClass("", MultiWebViewPlaceholder1Activity::class.java),
            ActivityClass("", MultiWebViewPlaceholder2Activity::class.java),
            ActivityClass("", MultiWebViewPlaceholder3Activity::class.java),
            ActivityClass("", MultiWebViewPlaceholder4Activity::class.java),
            ActivityClass("", MultiWebViewPlaceholder5Activity::class.java),
        )
        val controllerMap = mutableMapOf<Mmid, MultiWebViewController>()
        /**获取当前的controller, 只能给nativeUI 使用，因为他们是和mwebview绑定在一起的*/
        fun getCurrentWebViewController(mmid: Mmid): MultiWebViewController? {
            return controllerMap[mmid]
        }
    }

    override suspend fun _bootstrap(bootstrapContext: BootstrapContext) {
        /// nativeui 与 mwebview 是伴生关系
        bootstrapContext.dns.bootstrap("nativeui.sys.dweb")

        // 打开webview
        val queryUrl = Query.string().required("url")
        val queryWebviewId = Query.string().required("webview_id")

        apiRouting = routes(
            // 打开一个 webview 作为窗口
            "/open" bind Method.GET to defineHandler { request, ipc ->
                val url = queryUrl(request)
                val remoteMm = ipc.asRemoteInstance()
                    ?: throw Exception("mwebview.sys.dweb/open should be call by locale")
                val viewItem = openDwebView(remoteMm, url)
                Response(Status.OK).body(viewItem.webviewId)
            },
            // 关闭指定 webview 窗口
            "/close" bind Method.GET to defineHandler { request, ipc ->
                val webviewId = queryWebviewId(request)
                val remoteMmid = ipc.remote.mmid

                closeDwebView(remoteMmid, webviewId)
            },
            // 界面没有关闭，用于重新唤醒
            "/activate" bind Method.GET to defineHandler { request, ipc ->
                val remoteMmid = ipc.remote.mmid
                val webViewId = queryWebviewId(request)
                debugMultiWebView("REOPEN-WEBVIEW", "remote-mmid: $remoteMmid==>$webViewId")
                activityClassList.find { it.mmid == remoteMmid }?.let { activityClass ->
                    App.startActivity(activityClass.ctor) { intent ->
                        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_DOCUMENT)
                        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)

                        val b = Bundle()
                        b.putString("mmid", remoteMmid)
                        intent.putExtras(b)
                    }
                }
                return@defineHandler Response(Status.OK).body(webViewId)
            },
        )
    }

    override suspend fun _shutdown() {
        apiRouting = null
    }


    override fun openActivity(remoteMmid: Mmid){
        val flags = mutableListOf<Int>(Intent.FLAG_ACTIVITY_NEW_DOCUMENT)
        val activityClass = activityClassList.find { it.mmid == remoteMmid } ?:
        // 如果没有，从第一个挪出来，放到最后一个，并将至付给 remoteMmid
        activityClassList.removeAt(0).also {
            it.mmid = remoteMmid
            activityClassList.add(it)
            flags.add(Intent.FLAG_ACTIVITY_MULTIPLE_TASK)
        }
        App.startActivity(activityClass.ctor) { intent ->
            // intent.addFlags(Intent.FLAG_ACTIVITY_MULTIPLE_TASK)
            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_DOCUMENT)
//            intent.addFlags(Intent.FLAG_ACTIVITY_BROUGHT_TO_FRONT)
            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)

            val b = Bundle();
            b.putString("mmid", remoteMmid);
            intent.putExtras(b);
        }
    }


    private suspend fun openDwebView(
        remoteMm: MicroModule,
        url: String,
    ): MultiWebViewController.ViewItem {
        val remoteMmid = remoteMm.mmid
        debugMultiWebView("OPEN-WEBVIEW", "remote-mmid: $remoteMmid / url:$url")
        val controller = controllerMap.getOrPut(remoteMmid) {
            MultiWebViewController(
                remoteMmid,
                this,
                remoteMm,
            )
        }

        openActivity(remoteMmid)
        activitySignal.emit(Pair(remoteMmid, controller.waitActivityCreated()))
        return controller.openWebView(url)
    }

    private suspend fun closeDwebView(remoteMmid: String, webviewId: String): Boolean {
        onDestroySignal.emit(remoteMmid)
        return controllerMap[remoteMmid]?.closeWebView(webviewId) ?: false
    }

}

File: next/dweb-browser/src/Platforms/iOS/MicroModule/Sys/Mwebview/MultiWebViewNMM.cs
using DwebBrowser.MicroService;
using DwebBrowser.MicroService.Core;
using DwebBrowser.MicroService.Message;
using System.Net;
using UIKit;

#nullable enable

namespace DwebBrowser.MicroService.Sys.Mwebview;

public class MultiWebViewNMM : IOSNativeMicroModule
{
    static Debugger Console = new Debugger("MultiWebViewNMM");

    public MultiWebViewNMM() : base("mwebview.sys.dweb")
    {
    }

    private static Dictionary<Mmid, MultiWebViewController> s_controllerMap = new();
    private static MultiWebViewController? s_getCurrentWebViewController(Mmid mmid) => s_controllerMap.GetValueOrDefault(mmid);

    protected override async Task _bootstrapAsync(IBootstrapContext bootstrapContext)
    {
        /// nativeui 与 mwebview 是伴生关系
        await bootstrapContext.Dns.BootstrapAsync("nativeui.sys.dweb");

        // 打开一个webview作为窗口
        HttpRouter.AddRoute(IpcMethod.Get, "/open", async (request, ipc) =>
        {
            var url = request.QueryValidate<string>("url")!;
            var remoteMM = ipc.AsRemoteInstance() ?? throw new Exception("mwebview.sys.dweb/open should be call by locale");

            var viewItem = await _openDwebViewAsync(remoteMM, url);
            return new HttpResponseMessage(HttpStatusCode.OK).Also(it => it.Content = new StringContent(viewItem.webviewId));
        });

        // 关闭指定 webview 窗口
        HttpRouter.AddRoute(IpcMethod.Get, "/close", async (request, ipc) =>
        {
            var webviewId = request.QueryValidate<string>("webview_id")!;
            var remoteMmid = ipc!.Remote.Mmid;

            return await _closeDwebViewAsync(remoteMmid, webviewId);
        });

        // 界面没有关闭，用于重新唤醒
        HttpRouter.AddRoute(IpcMethod.Get, "/activate", async (request, ipc) =>
        {
            var remoteMmid = ipc.Remote.Mmid;
            var webViewId = request.QueryValidate<string>("webview_id")!;

            Console.Log("REOPEN-WEBVIEW", "remote-mmid: {0} ==> {1}", remoteMmid, webViewId);
            OpenActivity(remoteMmid);

            return new HttpResponseMessage(HttpStatusCode.OK).Also(it => it.Content = new StringContent(webViewId));
        });
    }

    public override void OpenActivity(string remoteMmid)
    {
        var vc = IOSWindow.RootViewController;
        var controller = s_controllerMap.GetValueOrDefault(remoteMmid);
        vc?.NavigationController?.PushViewController(controller!, true);
    }

    private async Task<MultiWebViewController.ViewItem> _openDwebViewAsync(MicroModule remoteMM, string url)
    {
        var remoteMmid = remoteMM.Mmid;
        Console.Log("OPEN-WEBVIEW", "remote-mmid: {0} / url: {1}", remoteMmid, url);
        var controller = s_controllerMap.GetValueOrPut(remoteMmid, () => new MultiWebViewController(remoteMmid, this, remoteMM));

        OpenActivity(remoteMmid);
        await _OnActivityEmit(remoteMmid, controller);
        return await controller.OpenWebViewAsync(url);
    }

    private async Task<bool> _closeDwebViewAsync(string remoteMmid, string webviewId)
    {
        var controller = s_controllerMap.GetValueOrDefault(remoteMmid);

        if (controller is not null)
        {
            await controller.CloseWebViewAsync(webviewId);
            await controller.DismissViewControllerAsync(true);
        }

        return false;
    } 
}

